### PR TITLE
Document `Boost::math` oddities

### LINKED ISSUE #1098: Broken Package: Boost::math
I have an issue using Boost::math and I've minimized it pretty well.

Given a CMakeLists.txt of:

```cmake
cmake_minimum_required(VERSION 3.6)

include("cmake/HunterGate.cmake")
HunterGate(
    URL "https://github.com/ruslo/hunter/archive/v0.19.126.tar.gz"
    SHA1 "78fb48ee9e636adcadbc61a6895a338733388dc4"
)

project(tmp_test)

hunter_add_package(Boost COMPONENTS math)
find_package(Boost CONFIG REQUIRED math)
```

And running in a directory with only that above CMakeLists.txt file (and huntergate of course):

```zsh
╰─➤  rm -r ~/.hunter; cmake3 .
```

Returns [CMakeOutput.log](https://overminddl1.com/tmp/CMakeOutput.log) and the [stdout/stderr](https://overminddl1.com/tmp/log.txt).
Of which the most interesting parts (at least to me) are:

```zsh
[ 87%] Performing install step for 'Boost-math'
Performing configuration checks

    - 32-bit                   : no  (cached)
    - 64-bit                   : yes (cached)
    - arm                      : no  (cached)
    - mips1                    : no  (cached)
    - power                    : no  (cached)
    - sparc                    : no  (cached)
    - x86                      : yes (cached)
    - symlinks supported       : yes (cached)
    - gcc visibility           : yes (cached)
    - long double support      : yes (cached)

Component configuration:

    - atomic                   : not building
    - chrono                   : not building
    - container                : not building
    - context                  : not building
    - coroutine                : not building
    - date_time                : not building
    - exception                : not building
    - fiber                    : not building
    - filesystem               : not building
    - graph                    : not building
    - graph_parallel           : not building
    - iostreams                : not building
    - locale                   : not building
    - log                      : not building
    - math                     : building
    - metaparse                : not building
    - mpi                      : not building
    - program_options          : not building
    - python                   : not building
    - random                   : not building
    - regex                    : not building
    - serialization            : not building
    - signals                  : not building
    - stacktrace               : not building
    - system                   : not building
    - test                     : not building
    - thread                   : not building
    - timer                    : not building
    - type_erasure             : not building
    - wave                     : not building

[100%] Completed 'Boost-math'
[100%] Built target Boost-math
-- [hunter] Build step successful (dir: /home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Build/Boost/__math)
-- [hunter] Cache saved: /home/my_user/.hunter/_Base/Cache/raw/60f6432f9483b19605d5719834db3044f02f4b40.tar.bz2
CMake Error at /home/my_user/.hunter/_Base/Download/Hunter/0.19.126/78fb48e/Unpacked/cmake/find/FindBoost.cmake:1247 (message):
  Unable to find the requested Boost libraries.

  Boost version: 1.65.1

  Boost include path:
  /home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/include

  Could not find the following static Boost libraries:

          boost_math

  No Boost libraries were found.  You may need to set BOOST_LIBRARYDIR to the
  directory containing Boost libraries or BOOST_ROOT to the location of
  Boost.
Call Stack (most recent call first):
  /home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/cmake/Boost/BoostConfig.cmake:40 (find_package)
  CMakeLists.txt:12 (find_package)



[hunter ** INTERNAL **] Library not exists: Boost_MATH_LIBRARY_DEBUG-NOTFOUND
[hunter ** INTERNAL **] [Directory:/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/cmake/Boost]

------------------------------ WIKI -------------------------------
    https://github.com/ruslo/hunter/wiki/error.internal
-------------------------------------------------------------------

CMake Error at /home/my_user/.hunter/_Base/Download/Hunter/0.19.126/78fb48e/Unpacked/cmake/modules/hunter_wiki.cmake:12 (message):
Call Stack (most recent call first):
  /home/my_user/.hunter/_Base/Download/Hunter/0.19.126/78fb48e/Unpacked/cmake/modules/hunter_internal_error.cmake:13 (hunter_wiki)
  /home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/cmake/Boost/BoostConfig.cmake:88 (hunter_internal_error)
  CMakeLists.txt:12 (find_package)


CMake Error at CMakeLists.txt:12 (find_package):
  Found package configuration file:

    /home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/cmake/Boost/BoostConfig.cmake

  but it set Boost_FOUND to FALSE so package "Boost" is considered to be NOT
  FOUND.


-- Configuring incomplete, errors occurred!
```

And yet:

```zsh
/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/libboost_math_c99-mt-d.a
/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/libboost_math_c99-mt.a
/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/libboost_math_c99f-mt-d.a
/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/libboost_math_c99f-mt.a
/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/libboost_math_c99l-mt-d.a
/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/libboost_math_c99l-mt.a
/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/libboost_math_tr1-mt-d.a
/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/libboost_math_tr1-mt.a
/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/libboost_math_tr1f-mt-d.a
/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/libboost_math_tr1f-mt.a
/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/libboost_math_tr1l-mt-d.a
/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/libboost_math_tr1l-mt.a
/home/my_user/.hunter/_Base/Cellar/60f6432f9483b19605d5719834db3044f02f4b40/60f6432/raw/lib/libboost_math_tr1-mt.a
/home/my_user/.hunter/_Base/Cellar/60f6432f9483b19605d5719834db3044f02f4b40/60f6432/raw/lib/libboost_math_tr1f-mt.a
/home/my_user/.hunter/_Base/Cellar/60f6432f9483b19605d5719834db3044f02f4b40/60f6432/raw/lib/libboost_math_tr1l-mt.a
/home/my_user/.hunter/_Base/Cellar/60f6432f9483b19605d5719834db3044f02f4b40/60f6432/raw/lib/libboost_math_c99-mt.a
/home/my_user/.hunter/_Base/Cellar/60f6432f9483b19605d5719834db3044f02f4b40/60f6432/raw/lib/libboost_math_c99f-mt.a
/home/my_user/.hunter/_Base/Cellar/60f6432f9483b19605d5719834db3044f02f4b40/60f6432/raw/lib/libboost_math_c99l-mt.a
/home/my_user/.hunter/_Base/Cellar/60f6432f9483b19605d5719834db3044f02f4b40/60f6432/raw/lib/libboost_math_tr1-mt-d.a
/home/my_user/.hunter/_Base/Cellar/60f6432f9483b19605d5719834db3044f02f4b40/60f6432/raw/lib/libboost_math_tr1f-mt-d.a
/home/my_user/.hunter/_Base/Cellar/60f6432f9483b19605d5719834db3044f02f4b40/60f6432/raw/lib/libboost_math_tr1l-mt-d.a
/home/my_user/.hunter/_Base/Cellar/60f6432f9483b19605d5719834db3044f02f4b40/60f6432/raw/lib/libboost_math_c99-mt-d.a
/home/my_user/.hunter/_Base/Cellar/60f6432f9483b19605d5719834db3044f02f4b40/60f6432/raw/lib/libboost_math_c99f-mt-d.a
/home/my_user/.hunter/_Base/Cellar/60f6432f9483b19605d5719834db3044f02f4b40/60f6432/raw/lib/libboost_math_c99l-mt-d.a
```
And the headers and all are there as expected, including the Config file.  And as you can see in the actual log files, it did all compile and install properly.

Running the same command without removing `~/.hunter` just prints out this in a fresh directory with just the CMakeLists.txt file (and huntergate of course):

```zsh
─➤  rm -r _3rdParty CMakeFiles CMakeCache.txt; cmake3 .                                                                        1 ↵
-- The C compiler identification is GNU 4.8.5
-- The CXX compiler identification is GNU 4.8.5
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- [hunter] Calculating Toolchain-SHA1
-- [hunter] Calculating Config-SHA1
-- [hunter] HUNTER_ROOT: /home/my_user/.hunter
-- [hunter] [ Hunter-ID: 78fb48e | Toolchain-ID: a397f94 | Config-ID: 9d1abe4 ]
-- [hunter] BOOST_ROOT: /home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install (ver.: 1.65.1)
-- [hunter] BOOST_ROOT: /home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install (ver.: 1.65.1)
CMake Error at /home/my_user/.hunter/_Base/Download/Hunter/0.19.126/78fb48e/Unpacked/cmake/find/FindBoost.cmake:1247 (message):
  Unable to find the requested Boost libraries.

  Boost version: 1.65.1

  Boost include path:
  /home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/include

  Could not find the following static Boost libraries:

          boost_math

  No Boost libraries were found.  You may need to set BOOST_LIBRARYDIR to the
  directory containing Boost libraries or BOOST_ROOT to the location of
  Boost.
Call Stack (most recent call first):
  /home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/cmake/Boost/BoostConfig.cmake:40 (find_package)
  CMakeLists.txt:12 (find_package)



[hunter ** INTERNAL **] Library not exists: Boost_MATH_LIBRARY_DEBUG-NOTFOUND
[hunter ** INTERNAL **] [Directory:/home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/cmake/Boost]

------------------------------ WIKI -------------------------------
    https://github.com/ruslo/hunter/wiki/error.internal
-------------------------------------------------------------------

CMake Error at /home/my_user/.hunter/_Base/Download/Hunter/0.19.126/78fb48e/Unpacked/cmake/modules/hunter_wiki.cmake:12 (message):
Call Stack (most recent call first):
  /home/my_user/.hunter/_Base/Download/Hunter/0.19.126/78fb48e/Unpacked/cmake/modules/hunter_internal_error.cmake:13 (hunter_wiki)
  /home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/cmake/Boost/BoostConfig.cmake:88 (hunter_internal_error)
  CMakeLists.txt:12 (find_package)


CMake Error at CMakeLists.txt:12 (find_package):
  Found package configuration file:

    /home/my_user/.hunter/_Base/78fb48e/a397f94/9d1abe4/Install/lib/cmake/Boost/BoostConfig.cmake

  but it set Boost_FOUND to FALSE so package "Boost" is considered to be NOT
  FOUND.


-- Configuring incomplete, errors occurred!
See also "/home/my_user/tmp/test_projects/tmp/CMakeFiles/CMakeOutput.log".
```

My system is:

```zsh
─➤  uname -a
Linux servername 3.10.0-514.26.2.el7.x86_64 #1 SMP Fri Jun 30 05:26:04 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

### PR DESCRIPTION
**Pull request by OvermindDL1**
Fixes #1098 but I'm not quite happy with how the restructuredtext is formatted, is that really becoming italic and why...  Any ideas on how to fix that formatting?

### CODE REVIEW

File: docs/packages/pkg/Boost.rst [hunk 1]
```diff
@@ -42,7 +43,13 @@ List of components (other libraries are header-only):
 - ``iostreams``
 - ``locale``
 - ``log``
-- ``math``
+- ``math`` - Requires Special names for ``find_package`` and ``target_link_libraries``, choose one of:
```
Comment (ruslo) [inline docs/packages/pkg/Boost.rst]: This is a list of components that can be used in `hunter_add_package(Boost COMPONENTS ...)` so it doesn't belong here. For the `find_package(Boost CONFIG REQUIRED ...)` you can add code snippet similar to Qt:
* https://docs.hunter.sh/en/latest/packages/pkg/Qt.html#usage

File: examples/Boost-math/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#find_package(Boost CONFIG REQUIRED math_c99)
+#find_package(Boost CONFIG REQUIRED math_c99f)
+#find_package(Boost CONFIG REQUIRED math_c99l)
+find_package(Boost CONFIG REQUIRED math_tr1)
+#find_package(Boost CONFIG REQUIRED math_tr1f)
+#find_package(Boost CONFIG REQUIRED math_tr1l)
+add_executable(foo foo.cpp)
+target_link_libraries(foo Boost::math_tr1)
+
+# Compatibility mode
```
Comment (ruslo) [inline examples/Boost-math/CMakeLists.txt]: We are testing it in other examples. Let's remove this.

File: examples/Boost-math/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+# * https://github.com/hunter-packages/gate
+include("../common.cmake")
+
+project(download-boost)
+
+# download boost
+hunter_add_package(Boost COMPONENTS math)
+
+# now boost can be used
+# you need to choose which form of Boost::math you want
```
Comment (ruslo) [inline examples/Boost-math/CMakeLists.txt]: Can we use all of them simultaneously?

**Comment by ruslo** [comment]
> I'm not quite happy with how the restructuredtext is formatted, is that really becoming italic and why... Any ideas on how to fix that formatting?

Actually it's bold for me, not italic. Are you talking about GitHub rendering or readthedocs (sphinx)? You can find documentation build instructions here:
* https://docs.hunter.sh/en/latest/creating-new/create/cmake.html#add-documentation

File: docs/packages/pkg/Boost.rst (hunk 1 above)
Comment (OvermindDL1) [reply docs/packages/pkg/Boost.rst]: Shall make a new section for just `math` then, was trying not to expand it that far as it increases verbosity, but yeah it is probably required, shall do.

File: examples/Boost-math/CMakeLists.txt (hunk 2 above)
Comment (OvermindDL1) [reply examples/Boost-math/CMakeLists.txt]: Gladly.  ^.^

**Comment by OvermindDL1** [comment]
> Actually it's bold for me, not italic. Are you talking about GitHub rendering or readthedocs (sphinx)? You can find documentation build instructions here:

Yeah the github renderer.  I'm on an entirely terminal system here currently so I cannot see generated docs locally, I have to git push to github and check it, say, on my phone to see how it looks.

File: examples/Boost-math/CMakeLists.txt (hunk 3 above)
Comment (ruslo) [reply examples/Boost-math/CMakeLists.txt]: Let's link all of them. By this we will tests that all `Boost::math_*` targets are valid.

**Comment by ruslo** [comment]
> Yeah the github renderer

They work differrently and will not look the same.

**Comment by OvermindDL1** [comment]
> They work differrently and will not look the same.

Ah, good to know that the restructured text format is so globally consistent...  >.>

**Comment by OvermindDL1** [comment]
What other changes are needed?  I have time today to do so now.  :-)

File: docs/packages/pkg/Boost.rst [hunk 4]
```diff
@@ -89,6 +90,44 @@ For example:
...
+    )
+
+Or implicitly:
+
+.. code-block:: cmake
+
+    hunter_add_package(Boost COMPONENTS math)
+    find_package(Boost CONFIG REQUIRED math_c99 math_c99f math_c99l math_tr1 math_tr1f math_tr1l)
+    include_directories(${Boost_INCLUDE_DIRS})
+    target_link_libraries(... ${Boost_LIBRARIES})
```
Comment (ruslo) [inline docs/packages/pkg/Boost.rst]: `Boost_LIBRARIES` is a list of full paths, it's not recommended to use it.

File: examples/Boost-math/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+# Emulate HunterGate:
+# * https://github.com/hunter-packages/gate
+include("../common.cmake")
+
+project(download-boost)
+
+# download boost
+hunter_add_package(Boost COMPONENTS math)
+
+# now boost::math can be used
```
Comment (ruslo) [inline examples/Boost-math/CMakeLists.txt]: `now boost::math` -> `now boost::math_* targets`

File: docs/packages/pkg/Boost.rst [hunk 6]
```diff
@@ -89,6 +90,44 @@ For example:
 -  `boost.iostreams
    options <http://www.boost.org/doc/libs/1_57_0/libs/iostreams/doc/index.html?path=7>`__
 
+Boost::Math
+-----------
+
+When using Boost Math you will need to link in the libraries, however these are not named ``math`` but
+rather are individual based on what you need to link it, the easist of which is to link in all parts:
```
Comment (ruslo) [inline docs/packages/pkg/Boost.rst]: `s,easist,easiest,`
* https://travis-ci.org/ruslo/hunter/builds/291658459#L1922

File: docs/packages/pkg/Boost.rst [hunk 7]
```diff
@@ -89,6 +90,44 @@ For example:
 -  `boost.iostreams
    options <http://www.boost.org/doc/libs/1_57_0/libs/iostreams/doc/index.html?path=7>`__
 
+Boost::Math
```
Comment (ruslo) [inline docs/packages/pkg/Boost.rst]: This is a little bit confusing. There is no imported library `Boost::Math`, C++ namespace is `boost::`. Let it be just `Math` then.

File: docs/packages/pkg/Boost.rst [hunk 8]
```diff
@@ -89,6 +90,44 @@ For example:
...
+-----------
+
+When using Boost Math you will need to link in the libraries, however these are not named ``math`` but
+rather are individual based on what you need to link it, the easist of which is to link in all parts:
+
+.. code-block:: cmake
+
+    hunter_add_package(Boost COMPONENTS math)
+    find_package(Boost CONFIG REQUIRED math_c99 math_c99f math_c99l math_tr1 math_tr1f math_tr1l)
+    include_directories(${Boost_INCLUDE_DIRS})
```
Comment (ruslo) [inline docs/packages/pkg/Boost.rst]: Not needed

File: docs/packages/pkg/Boost.rst (hunk 4 above)
Comment (OvermindDL1) [reply docs/packages/pkg/Boost.rst]: Aren't CMake build's non-relocateable anyway?  Plus `${Boost_INCLUDE_DIRS}` would make it non-relocateable too?

**Comment by OvermindDL1** [comment]
Any other changes?  :-)

**Comment by ruslo** [comment]
Testing with new math example:
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.2615
* https://travis-ci.org/ingenue/hunter/builds/291755892?utm_source=github_status&utm_medium=notification

**Comment by ruslo** [comment]
I've removed OSX Boost.math testing (we have Visual Studio and Linux) because of the timeout, don't know what is the problem. My local test worked fine.

If you want to check it and add new toolchain take a look at this documents and open pull request:
* https://docs.hunter.sh/en/latest/creating-new/update.html#add-toolchains
* https://docs.hunter.sh/en/latest/creating-new/create/cmake.html#ci-testing

Uploading:
* https://travis-ci.org/ingenue/hunter/builds/292422010?utm_source=github_status&utm_medium=notification
* https://ci.appveyor.com/project/ingenue/hunter/build/1.0.2629

**Comment by ruslo** [comment]
Released: https://github.com/ruslo/hunter/releases/tag/v0.19.134

**Pull request merged**